File: lib/aubo_sdk-0.27.1-rc.4-Windows_AMD64+b46f170/share/example/c++/example_servoj3.cpp
#include <cstring>
#include <fstream>
#include <math.h>
#ifdef WIN32
#include <windows.h>
#endif

#include "aubo_sdk/rpc.h"
#include "trajectory_io.h"

using namespace arcs::aubo_sdk;
using namespace arcs::common_interface;

#ifndef M_PI
#define M_PI 3.14159265358979323846
#endif
#define EMBEDDED

// 实现阻塞功能: 当机械臂运动到目标路点时，程序再往下执行
int waitArrival(RobotInterfacePtr impl)
{
    const int max_retry_count = 5;
    int cnt = 0;

    // 接口调用: 获取当前的运动指令 ID
    int exec_id = impl->getMotionControl()->getExecId();

    // 等待机械臂开始运动
    while (exec_id == -1) {
        if (cnt++ > max_retry_count) {
            return -1;
        }
        std::this_thread::sleep_for(std::chrono::milliseconds(50));
        exec_id = impl->getMotionControl()->getExecId();
    }

    // 等待机械臂动作完成
    while (impl->getMotionControl()->getExecId() != -1) {
        std::this_thread::sleep_for(std::chrono::milliseconds(50));
    }

    return 0;
}

template <typename T>
inline std::ostream &operator<<(std::ostream &os, const std::vector<T> &list)
{
    for (size_t i = 0; i < list.size(); i++) {
        os << list.at(i);
        if (i != (list.size() - 1)) {
            os << ",";
        }
    }
    return os;
}

int waitServoJointComplete(RobotInterfacePtr impl, int timeout_ms = 10000)
{
    auto start_time = std::chrono::steady_clock::now();

    while (impl->getMotionControl()->getMotionLeftTime(0) != 0) {
        auto current_time = std::chrono::steady_clock::now();
        auto elapsed_ms = std::chrono::duration_cast<std::chrono::milliseconds>(
                              current_time - start_time)
                              .count();

        if (elapsed_ms >= timeout_ms) {
            std::cout << "Warning: WaitServoJointComplete timeout after "
                      << elapsed_ms << " ms" << std::endl;
            return 1;
        }

        std::this_thread::sleep_for(std::chrono::milliseconds(5));
    }

    while (!impl->getRobotState()->isSteady()) {
        std::this_thread::sleep_for(std::chrono::milliseconds(5));
    }

    return 0;
}

/**
 * 测试1:
 * 开启力控 采用servoj跟踪一个轨迹,目标点下发时间间隔5ms
 */
int exampleServoj1(RpcClientPtr cli)
{
    auto robot_name = cli->getRobotNames().front();
#ifdef EMBEDDED
    // 内置传感器
    std::vector<double> sensor_pose = { 0, 0, 0, 0, 0, 0 };
    cli->getRobotInterface(robot_name)
        ->getRobotConfig()
        ->selectTcpForceSensor("embedded");
#else
    // 外置坤维传感器
    std::vector<double> sensor_pose = { 0, 0, 0.047, 0, 0, 0 };
    cli->getRobotInterface(robot_name)
        ->getRobotConfig()
        ->selectTcpForceSensor("kw_ftsensor");
#endif

    // 设置传感器安装位姿
    cli->getRobotInterface(robot_name)
        ->getRobotConfig()
        ->setTcpForceSensorPose(sensor_pose);
    // 设置TCP偏移
    std::vector<double> tcp_pose = { 0, 0, 0.0, 0, 0, 0 };
    cli->getRobotInterface(robot_name)
        ->getRobotConfig()
        ->setTcpOffset(tcp_pose);

    double mass = 0.0;
    std::vector<double> com = { 0.0, 0.0, 0.0 };
    // 设置负载
    cli->getRobotInterface(robot_name)
        ->getRobotConfig()
        ->setPayload(mass, com, { 0. }, { 0. });
    // 力传感器偏移需要根据实际情况设置
    std::vector<double> force_offset = cli->getRobotInterface(robot_name)
                                           ->getRobotState()
                                           ->getTcpForceSensors();

    // 设置力传感器偏移
    cli->getRobotInterface(robot_name)
        ->getRobotConfig()
        ->setTcpForceOffset(force_offset);

    std::vector<double> admittance_m = { 25.0, 25.0, 25.0, 2.0, 2.0, 2.0 };
    std::vector<double> admittance_d = {
        300.0, 300.0, 300.0, 12.0, 12.0, 12.0
    };
    std::vector<double> admittance_k = { 0.0, 0.0, 0.0, 0.0, 0.0, 0.0 };

    cli->getRobotInterface("rob1")->getForceControl()->setDynamicModel(
        admittance_m, admittance_d, admittance_k);

    // 力控开启方向
    std::vector<bool> compliance = { false, false, true, false, false, false };
    // 目标力
    std::vector<double> target_wrench{ 0.0, 0.0, 5.0, 0.0, 0.0, 0.0 };
    std::vector<double> speed_limits(6, 2.0);

    std::vector<double> feature = { 0, 0, 0, 0, 0, 0 };
    TaskFrameType frame_type = TaskFrameType::TOOL_FORCE;
    cli->getRobotInterface(robot_name)
        ->getForceControl()
        ->setTargetForce(feature, compliance, target_wrench, speed_limits,
                         frame_type);

    // 读取轨迹文件
    auto filename = "../trajs/record6.offt";
    TrajectoryIo input(filename);

    // 尝试打开轨迹文件，如果无法打开，直接返回
    if (!input.open()) {
        return 0;
    }

    // 解析轨迹数据
    auto traj = input.parse();

    // 检查轨迹文件中是否有路点，
    // 如果数量为 0，输出错误消息并返回
    auto traj_sz = traj.size();
    if (traj_sz == 0) {
        std::cerr << "轨迹文件中的路点数量为0." << std::endl;
        return 0;
    }

    // 接口调用: 获取机器人的名字
    auto robot_interface = cli->getRobotInterface(robot_name);

    // 接口调用: 设置机械臂的速度比率
    robot_interface->getMotionControl()->setSpeedFraction(1);

    // 接口调用: 关节运动到轨迹中的第一个点，否则容易引起较大超调
    robot_interface->getMotionControl()->moveJoint(traj[99], 80 * (M_PI / 180),
                                                   60 * (M_PI / 180), 0, 0);

    // 阻塞
    int ret = waitArrival(robot_interface);
    if (ret == 0) {
        std::cout << "关节运动到轨迹文件中的第一个路点成功" << std::endl;
    } else {
        std::cout << "关节运动到轨迹文件中的第一个路点失败" << std::endl;
    }
    robot_interface->getForceControl()->fcEnable();

    // 接口调用: 开启 servo 模式 7
    cli->getRobotInterface(robot_name)
        ->getMotionControl()
        ->setServoModeSelect(7);
    // 等待进入 servo 模式
    int i = 0;
    while (cli->getRobotInterface(robot_name)
               ->getMotionControl()
               ->getServoModeSelect() != 7) {
        if (i++ > 20) {
            std::cout << "Servo 模式使能失败! 当前的 Servo 模式是 "
                      << cli->getRobotInterface(robot_name)
                             ->getMotionControl()
                             ->getServoModeSelect()
                      << std::endl;
            return -1;
        }
        std::this_thread::sleep_for(std::chrono::milliseconds(7));
    }

    for (size_t i = 100; i < traj.size(); i++) {
        // 接口调用: 关节伺服运动

        while (cli->getRobotInterface(robot_name)
                   ->getMotionControl()
                   ->servoJoint(traj[i], 0.1, 0.2, 0.05, 0.1, 200) == 2) {
            // 缓存满了，需要重新下发
            // std::this_thread::sleep_for(std::chrono::milliseconds(5));
            std::cout << "Buffer is full，i: " << i << std::endl;
        }

        // std::this_thread::sleep_for(std::chrono::milliseconds(5));
    }

    // 等待运动结束
    int wait_servo = waitServoJointComplete(robot_interface);

    if (wait_servo == 0) {
        std::cout << "Servoj 运动结束" << std::endl;
    }

    // 接口调用: 关闭 servo 模式  7
    cli->getRobotInterface(robot_name)
        ->getMotionControl()
        ->setServoModeSelect(0);

    // 等待结束 servo 模式7
    while (cli->getRobotInterface(robot_name)
               ->getMotionControl()
               ->getServoModeSelect() != 0) {
        if (i++ > 20) {
            std::cout << "Servo 模式失能失败! 当前的 Servo 模式是 "
                      << cli->getRobotInterface(robot_name)
                             ->getMotionControl()
                             ->getServoModeSelect()
                      << std::endl;
            return -1;
        }
        std::this_thread::sleep_for(std::chrono::milliseconds(5));
    }
    cli->getRobotInterface("rob1")->getForceControl()->fcDisable();

    return 0;
}

#define LOCAL_IP "192.168.3.10"
/**
 * servoj功能使用步骤:
 * 1、采用实时系统测试servoj功能
 * 2、孤立一个CPU，保证该CPU没有其他任务
 * 3、将该进程设置为实时进程，优先级设置为最大
 * 4、绑定CPU，将该实时进程绑定到第二步孤立的CPU上
 */
int main(void)
{
#ifdef WIN32
    // 将Windows控制台输出代码页设置为 UTF-8
    SetConsoleOutputCP(CP_UTF8);
#endif
#ifndef _WIN32
    // 实时优先级最大值、最小值
    int sched_max = sched_get_priority_max(SCHED_FIFO);

    // 设置实时调度策略及优先级
    struct sched_param sParam;
    sParam.sched_priority = sched_max;
    sched_setscheduler(0, SCHED_FIFO, &sParam);
    auto i_schedFlag = sched_getscheduler(0);
    printf("设置调度策略 = [%d]\n", i_schedFlag);

    // 绑定CPU
    cpu_set_t cpuset;
    CPU_ZERO(&cpuset);
    CPU_SET(1, &cpuset);

    // bind process to processor 0
    if (sched_setaffinity(0, sizeof(cpuset), &cpuset) < 0) {
        perror("Sched_setaffinity fail!");
    }
#endif

    auto rpc_cli = std::make_shared<RpcClient>();
    // 接口调用: 设置 RPC 超时
    rpc_cli->setRequestTimeout(1000);
    // 接口调用: 连接到 RPC 服务
    rpc_cli->connect(LOCAL_IP, 30004);
    // 接口调用: 登录
    rpc_cli->login("aubo", "123456");

    exampleServoj1(rpc_cli);

    // 接口调用: RPC 退出登录
    rpc_cli->logout();
    // 接口调用: RPC 断开连接
    rpc_cli->disconnect();

    return 0;
}

File: lib/aubo_sdk-0.27.1-rc.4-Windows_AMD64+b46f170/share/example/python/example_servoj3.py
#! /usr/bin/env python
# coding=utf-8

"""
功能：机械臂 servoJ 叠加力控示例

步骤:
第一步: 设置 RPC 超时、连接 RPC 服务、机械臂登录
第二步: servoJ 叠加力控示例
第三步: RPC 退出登录、断开连接
"""

import time
import math
import pyaubo_sdk
import csv
from typing import List, Sequence, Tuple, Any, Optional

robot_ip = "127.0.0.1"  # 服务器 IP 地址
robot_port = 30004  # 端口号
M_PI = 3.14159265358979323846
tcp_offset = [0.0, 0.0, 0.0, 0.0, 0.0, 0.0]
mass = 0.0
com = [0.0, 0.0, 0.0]
t = 0.05
robot_rpc_client = pyaubo_sdk.RpcClient()
EMBEDDED: bool = True # 是否使用机械臂内置力传感器
SENSOR_VENDOR: str = "kw_ftsensor"  # 外置力传感器厂商名称


# 阻塞
def wait_arrival(robot_interface):
    max_retry_count = 5
    cnt = 0

    # 接口调用: 获取当前的运动指令 ID
    exec_id = robot_interface.getMotionControl().getExecId()

    # 等待机械臂开始运动
    while exec_id == -1:
        if cnt > max_retry_count:
            return -1
        time.sleep(0.05)
        cnt += 1
        exec_id = robot_interface.getMotionControl().getExecId()

    # 等待机械臂运动完成
    while robot_interface.getMotionControl().getExecId() != -1:
        time.sleep(0.05)

    return 0

def calculate_distance(p1: Sequence[float], p2: Optional[Sequence[float]] = None) -> float:
    """
    计算两点之间的欧氏距离，默认第二点为原点
    """
    if p2 is None:
        p2 = (0.0,) * 6
    if len(p1) != 6 or len(p2) != 6:
        return 0.0
    return math.sqrt(sum((p1[i] - p2[i]) ** 2 for i in range(3)))

def waitServoJointComplete(cli: Any, timeout_ms=10000) -> int:
    # 接口调用：获取机器人的名字
    robot_name = cli.getRobotNames()[0]
    robot_interface = cli.getRobotInterface(robot_name)
    # 记录开始时间
    start_time = time.perf_counter()

    # 第一步：等待运动剩余时间为0（循环检查getMotionLeftTime(0)）
    while robot_interface.getMotionControl().getMotionLeftTime(0) != 0:
        # 计算已耗时（毫秒）
        current_time = time.perf_counter()
        elapsed_ms = (current_time - start_time) * 1000
        # 检查是否超时
        if elapsed_ms >= timeout_ms:
            print(f"Warning: WaitServoJointComplete timeout after {elapsed_ms:.0f} ms")
            return 1
        # 每5ms检查一次
        time.sleep(0.005)
    # 第二步：等待机械臂进入稳定状态（isSteady()为True）
    while not robot_interface.getRobotState().isSteady():
        time.sleep(0.005)

    return 0

def switch_servoj_mode(cli: Any, mode: int) -> int:
    # 接口调用：获取机器人的名字
    robot_name = cli.getRobotNames()[0]
    robot_interface = cli.getRobotInterface(robot_name)
    mc = robot_interface.getMotionControl()
    mc.setServoModeSelect(mode)
    i = 0
    while mc.getServoModeSelect() != mode:
        i = i + 1
        if i > 20:
            print("Servo Mode is ", mc.getServoModeSelect())
            return -1
        time.sleep(0.007)
    return 0

def fc_enable(cli: Any)->int:
    # 接口调用：获取机器人的名字
    robot_name = cli.getRobotNames()[0]
    robot = cli.getRobotInterface(robot_name)
    # 开启力控
    ret = robot.getForceControl().fcEnable()
    if ret != 0:
        print("力控开启失败, 错误码: ", ret)
        return -1

def fc_disable(cli: Any)->int:
    # 接口调用：获取机器人的名字
    robot_name = cli.getRobotNames()[0]
    robot = cli.getRobotInterface(robot_name)
    # 关闭力控
    ret = robot.getForceControl().fcDisable()
    if ret != 0:
        print("力控关闭失败, 错误码: ", ret)
        return -1

def tcp_sensor_test(cli: Any, csv_path: str = "force.csv"):
    # 接口调用：获取机器人的名字
    robot_name = cli.getRobotNames()[0]

    # 选择力传感器类型
    if EMBEDDED:
        sensor_pose = [0, 0, 0, 0, 0, 0]
        cli.getRobotInterface(robot_name) \
            .getRobotConfig() \
            .selectTcpForceSensor("embedded")
    else:
        sensor_pose = [0, 0, 0.047, 0, 0, 0]
        cli.getRobotInterface(robot_name) \
            .getRobotConfig() \
            .selectTcpForceSensor(SENSOR_VENDOR)

    # 写入 CSV
    with open(csv_path, "a", newline="") as f:
        writer = csv.writer(f)

        # 表头
        writer.writerow(["Fx", "Fy", "Fz", "Tx", "Ty", "Tz"])
        f.flush()

        while True:
            sensor_data = cli.getRobotInterface(robot_name) \
                             .getRobotState() \
                             .getTcpForceSensors()

            print("--------------------------------------")
            for i, v in enumerate(sensor_data):
                print(f"获取TCP力传感器读数: {i + 1}: {v}")
            print("--------------------------------------")

            writer.writerow(list(sensor_data))
            f.flush()

            time.sleep(0.005)

# --------- 力传感器标定 ----------
def tcp_sensor_calibration(cli: Any, joints: List[List[float]]) -> Tuple[List[float], List[float], float]:
    robot_name = cli.getRobotNames()[0]
    robot_interface = cli.getRobotInterface(robot_name)

    print("goto p0")
    robot_interface.getMotionControl().moveJoint(
        joints[0],
        10.0 * (math.pi / 180.0),
        5.0 * (math.pi / 180.0),
        0, 0
    )
    wait_arrival(robot_interface)
    time.sleep(1.0)

    q1 = robot_interface.getRobotState().getJointPositions()
    tcp_force1 = robot_interface.getRobotState().getTcpForceSensors()

    print("goto p1")
    robot_interface.getMotionControl().moveJoint(
        joints[1],
        20.0 * (math.pi / 180.0),
        10.0 * (math.pi / 180.0),
        0, 0
    )
    wait_arrival(robot_interface)
    time.sleep(1.0)

    q2 = robot_interface.getRobotState().getJointPositions()
    tcp_force2 = robot_interface.getRobotState().getTcpForceSensors()

    print("goto p2")
    robot_interface.getMotionControl().moveJoint(
        joints[2],
        20.0 * (math.pi / 180.0),
        10.0 * (math.pi / 180.0),
        0, 0
    )
    wait_arrival(robot_interface)
    time.sleep(1.0)

    q3 = robot_interface.getRobotState().getJointPositions()
    tcp_force3 = robot_interface.getRobotState().getTcpForceSensors()

    pose1 = robot_interface.getRobotAlgorithm().forwardKinematics(q1)
    pose2 = robot_interface.getRobotAlgorithm().forwardKinematics(q2)
    pose3 = robot_interface.getRobotAlgorithm().forwardKinematics(q3)

    calib_forces = [list(tcp_force1), list(tcp_force2), list(tcp_force3)]
    calib_poses = [list(pose1[0]), list(pose2[0]), list(pose3[0])]

    # 三点标定接口
    result = robot_interface.getRobotAlgorithm().calibrateTcpForceSensor(calib_forces, calib_poses)

    return result

def exampleServoj1(cli: Any):
    robot_name = cli.getRobotNames()[0]  # 接口调用: 获取机器人的名字
    robot = cli.getRobotInterface(robot_name)


    # 选择力传感器类型
    if EMBEDDED:
        sensor_pose = [0, 0, 0, 0, 0, 0]
        robot.getRobotConfig().selectTcpForceSensor("embedded")
    else:
        sensor_pose = [0, 0, 0.047, 0, 0, 0]
        robot.getRobotConfig().selectTcpForceSensor(SENSOR_VENDOR)

    # 设置传感器安装位姿
    robot.getRobotConfig().setTcpForceSensorPose(sensor_pose)

    # 设置工具坐标系偏移
    robot.getRobotConfig().setTcpOffset(tcp_offset)

    # 设置负载
    robot.getRobotConfig().setPayload(mass, com, [0.], [0.])

    # 读取并设置力传感器偏移
    force_offset = robot.getRobotState().getTcpForceSensors()
    robot.getRobotConfig().setTcpForceOffset(force_offset)

    # 导纳模型
    admittance_m = [25.0, 25.0, 25.0, 2.0, 2.0, 2.0]
    admittance_d = [300.0, 300.0, 300.0, 12.0, 12.0, 12.0]
    admittance_k = [0.0, 0.0, 0.0, 0.0, 0.0, 0.0]
    robot.getForceControl().setDynamicModel(admittance_m, admittance_d, admittance_k)

    # 力控目标
    compliance = [False, False, True, False, False, False]
    target_wrench = [0.0, 0.0, 5.0, 0.0, 0.0, 0.0]
    speed_limits = [2.0] * 6
    feature = [0, 0, 0, 0, 0, 0]
    task_frame = pyaubo_sdk.TaskFrameType.TOOL_FORCE
    robot.getForceControl().setTargetForce(feature, compliance, target_wrench, speed_limits, task_frame)

    # 读取轨迹文件并加载轨迹点
    file = open('../trajs/record6.offt')
    traj = []
    for line in file:
        str_list = line.split(",")
        float_list = []
        for strs in str_list:
            float_list.append(float(strs))
        traj.append(float_list)

    traj_sz = len(traj)
    if traj_sz == 0:
        print("没有轨迹点")
    else:
        print("加载的轨迹点数量为: ", traj_sz)

    # 关节运动到第一个点
    # 当前位置要与轨迹中的第一个点一致，否则容易引起较大超调
    print("goto p1")
    mc = robot.getMotionControl()
    mc.moveJoint(traj[0], M_PI, M_PI, 0., 0.)
    wait_arrival(robot)

    # 开启力控
    fc_enable(cli)

    # 开启 servo 模式
    switch_servoj_mode(cli, 7)

    traj.remove(traj[0])
    for q in traj:
        ret = mc.servoJoint(q, 0.1, 0.2, t, 0.1, 200)
        while ret == 2:
            # print("queue full: ", ret)
            # time.sleep(0.005)
            ret = mc.servoJoint(q, 0.1, 0.2, t, 0.1, 200)
        if ret != 0:
            print("servoj error: ", ret)
            robot.getForceControl().fcDisable()
            return -1

        # 5ms 下发节拍
        # time.sleep(0.005)

    # 等待 servo 运动完成
    if (waitServoJointComplete(cli)== 0):
        print("Servoj motion complete")

    # 关闭 servo 模式
    switch_servoj_mode(cli, 0)
    # 关闭力控
    fc_disable(cli)

    print("Servoj end")
    return 0

if __name__ == '__main__':
    robot_rpc_client = pyaubo_sdk.RpcClient()
    robot_rpc_client.setRequestTimeout(1000)  # 接口调用: 设置 RPC 超时
    robot_rpc_client.connect(robot_ip, robot_port)  # 接口调用: 连接 RPC 服务
    if robot_rpc_client.hasConnected():
        print("RPC客户端连接成功!")
        robot_rpc_client.login("aubo", "123456")  # 接口调用: 登录
        if robot_rpc_client.hasLogined():
            print("RPC客户端登录成功!")
            exampleServoj1(robot_rpc_client)  # ServoJ示例
            robot_rpc_client.logout()  # 退出登录
            robot_rpc_client.disconnect()  # 断开连接
